### ISSUE TITLE
Gas Optimizations

### ISSUE THREAD

**Issue by 0xMilenov** [VIS-2/3adao-01-24#19]
| | Issue | Instances | Gas Savings |
| ---- | :--- | :--: | :--: |
| [[G-01](#g-01)] | Missing validation check in redeemReward | 2 | 6 |
| [[G-02](#g-02)] | Optimize gas by using `!= 0` over ` > 0` in `require()` for unsigned integers | 9 | 54 |
| [[G-03](#g-03)] | a = a + b consumes less gas than a += b for state variables | 7 | 112 |
| [[G-04](#g-04)] | Use assembly to validate msg.sender | 7 | 84 |
| [[G-05](#g-05)] | Store array length outside loops for efficiency | 3 | 9 |
| [[G-06](#g-06)] | Using assembly to check for zero can save gas | 7 | 49 |
| [[G-07](#g-07)] | Revert strings are less gas-efficient than custom errors | 14 | 700 |
| [[G-08](#g-08)] | Using a double if statement instead of a logical AND(&&) | 1 | 30 |
| [[G-09](#g-09)] | Use shift right/left instead of division/multiplication if possible | 2 | 44 |
| [[G-10](#g-10)] | Upgrade to a newer Solidity version. | 21 | N/A |
| [[G-11](#g-11)] | Increments/decrements can be unchecked in for-loops | 5 | 125 |
| [[G-12](#g-12)] | Separate require() conditions for gas efficiency | 2 | 6 |

### [G-01]<a name="g-01"></a> Missing validation check in redeemReward
**Context** [StabilityPool](https://github.com/3A-DAO/3a-borrowing-protocol/blob/816da9674b8e16802eb50478941dde60a2b8db7b/contracts/StabilityPool.sol#L183-L193)

**Impact** Users with a zero deposit are allowed to execute the function, potentially causing unnecessary gas expenditure.

**Description**
```solidity
function redeemReward() external {
    Snapshots memory snapshots = depositSnapshots[msg.sender];
    uint256 contributorDeposit = deposits[msg.sender];

    uint256 compoundedDeposit = _getCompoundedDepositFromSnapshots(contributorDeposit, snapshots);
    _redeemReward();
    _updateDepositAndSnapshots(msg.sender, compoundedDeposit);
}
```

**Recomendation**. Add validation check.
```diff
function redeemReward() external {
    Snapshots memory snapshots = depositSnapshots[msg.sender];
    uint256 contributorDeposit = deposits[msg.sender];
+  require(contributorDeposit > 0, "deposit-is-0")

    uint256 compoundedDeposit = _getCompoundedDepositFromSnapshots(contributorDeposit, snapshots);
    _redeemReward();
    _updateDepositAndSnapshots(msg.sender, compoundedDeposit);
}
```

### [G-02]<a name="g-02"></a> Optimize gas by using `!= 0` over ` > 0` in `require()` for unsigned integers
For contracts using Solidity versions up to `0.8.13`, there's a slight gas optimization to be had when using `!= 0` instead of ` > 0` in `require()` checks on unsigned integers.

When the optimizer is enabled, the `!= 0` check consumes 6 gas units less. Although it might appear that ` > 0` is more gas-efficient, this holds true only without the optimizer and outside of `require` statements.

For contracts where every gas unit matters, consider this optimization.

Reference: [Tweet by @gzeon](https://twitter.com/gzeon/status/1485428085885640706).

Ensure the Optimizer is enabled for maximal benefit.

```solidity
File: ~All files

     require(_amount  > 0, "amount-is-0");
```
```diff
+   require(_amount  != 0, "amount-is-0");
```
   
### [G-03]<a name="g-03"></a> a = a + b consumes less gas than a += b for state variables, except for arrays and mappings

```solidity
File: ~All files / Example with AuctionManager contract

}144:       _totalCollateralValue += _collateralValue;
}180:       _totalCollateralValue += _collateralValue;
```
```diff
+         _totalCollateralValue = _totalCollateralValue + _collateralValue;
+         _totalCollateralValue = _totalCollateralValue + _collateralValue;
```

[144](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L144), [180](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L180)

### [G-04]<a name="g-04"></a> Use assembly to validate msg.sender
We can use assembly to efficiently validate msg.sender with the least amount of opcodes necessary. For more details check the following report [here](https://code4rena.com/reports/2023-05-juicebox#g-06-use-assembly-to-validate-msgsender)

```solidity
File: contracts/AuctionManager.sol

}156:      require(msg.sender == address(liquidationRouter), "not-allowed");
```

[156](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L156)

```solidity
File: contracts/LiquidationRouter.sol

}84:       require(msg.sender == stabilityPool, "not-allowed");
}89:       require(msg.sender == lastResortLiquidation, "not-last-resort-liquidation");
```

[84](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/LiquidationRouter.sol#L84), [89](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/LiquidationRouter.sol#L89)

### [G-05]<a name="g-05"></a> Store array length outside loops for efficiency
When array length isn't cached, the Solidity compiler repeatedly reads it in every loop iteration.

For storage arrays, this means an added sload operation costing 100 extra gas for each subsequent iteration.

For memory arrays, it's an additional mload operation costing 3 extra gas post the first iteration.

```solidity
File: contracts/LiquidationRouter.sol

}125:      for (uint256 i = 0; i <  collateralSet.length(); i++) {
```

[125](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/LiquidationRouter.sol#L125)
```solidity
File: contracts/StabilityPool.sol

}276:       for (uint128 i = 0; i <  _snapshots.tokenToSArray.length; i++) {
}285:       for (uint128 i = 0; i <  nextTokensToSum_cached.length; i++) {
}424:       for (uint128 i = 0; i <  currentTokenToSArray.length; i++) {
}552:       for (uint128 j = 0; j <  _snapshots.tokenToSArray.length; j++) {
}561:       for (uint128 j = 0; j <  nextTokensToSum_cached.length; j++) {
}758:       for (uint256 i = 0; i <  _depositorCollateralGains.length; i++) {
```

[276](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/StabilityPool.sol#L276), [285](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/StabilityPool.sol#L285), [424](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/StabilityPool.sol#L424), [552](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/StabilityPool.sol#L552), [561](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/StabilityPool.sol#L561), [758](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/StabilityPool.sol#L758)
```solidity
File: contracts/Vault.sol

}177:       for (uint256 i = 0; i <  collateralSet.length(); i++) {
}385:       for (uint256 i = 0; i <  collateralSet.length(); i++) {
}408:       for (uint256 i = 0; i <  collateralSet.length(); i++) {
```

[177](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/Vault.sol#L177), [385](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/Vault.sol#L385), [408](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/Vault.sol#L408)

### [G-06]<a name="g-06"></a> Using assembly to check for zero can save gas
Using assembly to check for zero can save gas by allowing more direct access to the evm and reducing some of the overhead associated with high-level operations in solidity.

```solidity
File: ~All files / Example with BONQMath.sol

}67:       if (_minutes == 0) {
}77:       if (n % 2 == 0) {
```

[67](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/BONQMath.sol#L67), [77](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/BONQMath.sol#L77)

### [G-07]<a name="g-07"></a> Revert strings are less gas-efficient than custom errors
From Solidity v0.8.4 onward, custom errors have been introduced.

These errors provide a savings of approximately 50 gas each instance they're triggered, as they circumvent the need to allocate and keep the revert string.

Omitting these strings also conserves gas during deployment.

Furthermore, custom errors are versatile, applicable both inside and outside contracts, including in interfaces and libraries.

As stated in the Solidity blog:

With the introduction of Solidity v0.8.4, a streamlined and gas-efficient method has been provided to elucidate to users the reasons behind an operation's failure through custom errors.

Prior to this, although strings could be used to detail failure reasons (e.g., revert('Insufficient funds.');), they were notably costlier, especially in terms of deployment, and incorporating dynamic information into them was challenging.

It's advisable to transition all revert strings to custom errors in your solution, especially focusing on those that appear multiple times.

```solidity
File: ~All files

        require(...
```

### [G-08]<a name="g-08"></a> Using a double if statement instead of a logical AND(&&)
Using a double if statement instead of a logical AND (&&) can provide similar short-circuiting behavior whereas double if is slightly [more gas efficient](https://gist.github.com/DadeKuma/931ce6794a050201ec6544dbcc31316c).

```solidity
File: contracts/utils/linked-address-list.sol

}46:         if (_before && (_reference == address(0x0) || _reference == _list._first)) {
}53:         } else if (!_before && (_reference == address(0x0) || _reference == _list._last)) {
}91:         if (_element == _list._last && _element == _list._first) {
```

[46](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/linked-address-list.sol#L46), [53](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/linked-address-list.sol#L53), [91](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/linked-address-list.sol#L91)
### [G-09]<a name="g-09"></a> Use shift right/left instead of division/multiplication if possible
While the `DIV` / `MUL` opcode uses 5 gas, the `SHR` / `SHL` opcode only uses 3 gas. Furthermore, beware that Solidity's division operation also includes a division-by-0 prevention which is bypassed using shifting. Eventually, overflow checks are never performed for shift operations as they are done for arithmetic operations. Instead, the result is always truncated, so the calculation can be unchecked in Solidity version `0.8+`
- Use ` >> 1` instead of `/ 2`
- Use ` >> 2` instead of `/ 4`
- Use ` << 3` instead of `* 8`
- ...
- Use ` >> 5` instead of `/ 2^5 == / 32`
- Use ` << 6` instead of `* 2^6 == * 64`

TL;DR:
- Shifting left by N is like multiplying by 2^N (Each bits to the left is an increased power of 2)
- Shifting right by N is like dividing by 2^N (Each bits to the right is a decreased power of 2)

*Saves around 2 gas + 20 for unchecked per instance*

```solidity
File: contracts/utils/BONQMath.sol

}47:         decProd = (prod_xy + (DECIMAL_PRECISION / 2)) / DECIMAL_PRECISION;
}79:         n = n / 2;
}84:         n = (n - 1) / 2;
```

[47](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/BONQMath.sol#L47), [79](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/BONQMath.sol#L79), [84](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/BONQMath.sol#L84)
```solidity
File: contracts/utils/constants.sol

}21:     uint256 public constant PERCENT_05 = PERCENT / 2; // Represents 0.5%
```

[21](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/utils/constants.sol#L21)

### [G-10]<a name="g-10"></a> Upgrade to a newer Solidity version.
Using a more recent version of Solidity offers various benefits, including enhanced functionality, bug fixes, and potential gas savings. It ensures your smart contracts are more secure and compatible with the latest tools and libraries

```solidity
File: ~All files

}2: pragma solidity ^0.8.4;
```

### [G-11]<a name="g-11"></a> Increments/decrements can be unchecked in for-loops
In Solidity 0.8+, there's a default overflow check on unsigned integers. It's possible to uncheck this in for-loops and save some gas at each iteration, but at the cost of some code readability, as this uncheck cannot be made inline.

[ethereum/solidity#10695](https://github.com/ethereum/solidity/issues/10695)

The change would be:

```diff
- for (uint256 i; i <  numIterations; i++) {
+ for (uint256 i; i <  numIterations;) {
 // ...  
+   unchecked { ++i; }
}  
```

These save around **25 gas saved** per instance.

The same can be applied with decrements (which should use `break` when `i == 0`).

The risk of overflow is non-existent for `uint256`.

```solidity
File: ~All files / Example with AuctionManager.sol

}170:         for (uint256 i = 0; i <  _collateralsLength; i++) {
}259:         for (uint256 i = 0; i <  _collateralsLength; i++) {
}299:         for (uint256 i = 0; i <  _collateralsLength; i++) {
```

[170](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L170), [259](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L259), [299](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L299)

### [G-12]<a name="g-12"></a> Separate require() conditions for gas efficiency
It's beneficial to separate conditions in `require()` statements rather than using the `&&` operator.

As outlined in [this discussed issue](https://github.com/code-423n4/2022-01-xdefi-findings/issues/128), while there might be a slightly higher deployment gas cost initially, the overall gas savings in runtime calls makes this approach more cost-effective in the long run.

Implementing this change can result in a saving of approximately *3 gas per instance*.

```solidity
File: contracts/AuctionManager.sol

}232:      require(!_auction.auctionEnded && block.timestamp < = _auction.auctionEndTime, "auction-ended");
```

[232](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/AuctionManager.sol#L232)
```solidity
File: contracts/TokenToPriceFeed.sol

}92:       require(_mlr  >= 100 && _mlr < = _mcr, "MLR <  100 or MLR  > MCR");
```

[92](https://github.com/3A-DAO/3a-borrowing-protocol/tree/main/contracts/TokenToPriceFeed.sol#L92)

**Comment by 0xMilenov** [comment]
@antoniopel @CristianRicharte6  
I wanted to contribute to the protocol and make all these Gas Optimisations issues. 
All are free of charge to show you my respect.
